semantic-analysis/src/Analysis/Carrier/Statement/State.hs

```diff
@@ -0,0 +1,43 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Analysis.Carrier.Statement.State
+( -- * Messages
+  Message(..)
+  -- * Statement carrier
+, runStatement
+, StatementC(..)
+  -- * Statement effect
+, module Analysis.Effect.Statement
+) where
+
+import           Analysis.Effect.Statement hiding (Import)
+import qualified Analysis.Effect.Statement as S
+import           Control.Algebra
+import           Control.Carrier.State.Church
+import           Control.Monad.Fail as Fail
+import           Data.List.NonEmpty (NonEmpty)
+import           Data.Text (Text)
+
+-- Messages
+
+data Message
+  = Import (NonEmpty Text)
+  deriving (Eq, Ord, Show)
+
+
+-- Statement carrier
+
+runStatement :: ([Message] -> a -> m r) -> StatementC m a -> m r
+runStatement k (StatementC m) = runState (k . reverse) [] m
```

Comment:
`Statement` effects just track sequential, imperative directives, currently limited to imports. This collects them into `Message`s which are then returned via this handler.